File: javascript/src/local-skills.ts
import type { LocalSkillDescriptor } from "./types/skills.js";
import { functionCallOutputInput, pendingFunctionCalls } from "./local-functions.js";
import type { FunctionCallOutputInput } from "./types/input.js";
import type { AgentResponse, FunctionCallOutputItem } from "./types/responses.js";

const DEFAULT_FOCUS_MANIFEST_CHARS = 16000;
const DEFAULT_FOCUS_FILE_CHARS = 12000;

export interface LocalSkillDirectoryOptions {
  id?: string;
  name?: string;
  description?: string;
  max_manifest_chars?: number;
  metadata?: Record<string, unknown>;
}

export async function localSkillFromDirectory(
  rootDir: string,
  options: LocalSkillDirectoryOptions = {},
): Promise<LocalSkillDescriptor> {
  const fs = await import("node:fs/promises");
  const path = await import("node:path");
  const crypto = await import("node:crypto");
  const root = path.resolve(rootDir);
  const files = await walkFiles(fs, path, root, root);
  const hash = crypto.createHash("sha256");
  for (const rel of files.sort()) {
    hash.update(rel);
    hash.update("\0");
    hash.update(await fs.readFile(path.join(root, rel)));
    hash.update("\0");
  }
  const digest = "sha256:" + hash.digest("hex");
  const base = path.basename(root);
  const localSkillID = options.id ?? base;
  const { manifest, manifestTruncated } = await readLocalManifest(fs, path, root, options.max_manifest_chars ?? DEFAULT_FOCUS_MANIFEST_CHARS);
  return {
    local_skill_id: localSkillID,
    skill_ref: skillRefForLocal(localSkillID, digest),
    name: options.name ?? base,
    description: options.description,
    root_hint: root,
    digest,
    manifest,
    manifest_truncated: manifestTruncated,
    metadata: options.metadata,
  };
}

export function pendingLocalSkillCalls(response: AgentResponse): FunctionCallOutputItem[] {
  return pendingFunctionCalls(response).filter((call) => call.name === "skill_focus");
}

export async function runLocalSkillHandlers(
  response: AgentResponse,
  localSkills: LocalSkillDescriptor[],
): Promise<FunctionCallOutputInput[]> {
  const byRef = await localSkillsByRef(localSkills);
  const outputs: FunctionCallOutputInput[] = [];
  for (const call of pendingLocalSkillCalls(response)) {
    const args = call.arguments ? JSON.parse(call.arguments) as Record<string, unknown> : {};
    const payload = await focusLocalSkills(args, byRef);
    outputs.push(functionCallOutputInput(call.call_id, payload));
  }
  return outputs;
}

async function focusLocalSkills(
  args: Record<string, unknown>,
  byRef: Map<string, LocalSkillDescriptor>,
): Promise<Record<string, unknown>> {
  const maxManifestChars = manifestCharLimit(args);
  const maxFileChars = fileCharLimit(args);
  const items = Array.isArray(args.skills) ? args.skills : [];
  if (!Array.isArray(args.skills)) {
    return { data: [{ ok: false, error: { code: "invalid_skill_focus", message: "skills must be an array" } }] };
  }
  const data: Record<string, unknown>[] = [];
  for (const raw of items) {
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
      data.push({ ok: false, error: { code: "invalid_skill_focus", message: "skill item must be an object" } });
      continue;
    }
    const item = raw as Record<string, unknown>;
    const skillRef = String(item.skill_ref || "").trim();
    const result: Record<string, unknown> = { ok: false, skill_ref: skillRef, branch: "main" };
    const descriptor = byRef.get(skillRef);
    if (!descriptor) {
      result.error = { code: "skill_ref_not_found", message: "skill_ref was not registered with the SDK" };
      data.push(result);
      continue;
    }
    try {
      result.skill = await focusedLocalSkill(descriptor, maxManifestChars, pathsArg(item), maxFileChars, includeManifest(item));
      result.ok = true;
    } catch (error) {
      result.error = { code: "local_skill_focus_failed", message: error instanceof Error ? error.message : String(error) };
    }
    data.push(result);
  }
  return { data };
}

async function focusedLocalSkill(
  descriptor: LocalSkillDescriptor,
  maxManifestChars: number,
  paths: string[],
  maxFileChars: number,
  includeManifest: boolean,
): Promise<Record<string, unknown>> {
  const fs = await import("node:fs/promises");
  const path = await import("node:path");
  const { TextDecoder } = await import("node:util");
  const root = path.resolve(descriptor.root_hint || ".");
  const stat = await fs.stat(root);
  if (!stat.isDirectory()) {
    throw new Error(`local skill root is not a directory: ${root}`);
  }
  let manifest = "";
  let manifestTruncated = false;
  if (includeManifest) {
    try {
      const manifestBytes = await fs.readFile(path.join(root, "SKILL.md"));
      manifest = new TextDecoder("utf-8", { fatal: true }).decode(manifestBytes);
      const manifestChars = Array.from(manifest);
      if (maxManifestChars > 0 && manifestChars.length > maxManifestChars) {
        manifest = manifestChars.slice(0, maxManifestChars).join("");
        manifestTruncated = true;
      }
    } catch (error: any) {
      if (error?.code !== "ENOENT") {
        throw error;
      }
    }
  }
  const dirents = await fs.readdir(root, { withFileTypes: true });
  const entries = await Promise.all(
    dirents
      .filter((entry) => ![".git", "__pycache__", "node_modules"].includes(entry.name))
      .sort((a, b) => a.name.localeCompare(b.name))
      .map(async (entry) => {
        const full = path.join(root, entry.name);
        const entryStat = await fs.stat(full);
        return {
          path: entry.name,
          is_dir: entry.isDirectory(),
          size: entry.isDirectory() ? 0 : entryStat.size,
          modified_at: Math.floor(entryStat.mtimeMs / 1000),
        };
      }),
  );
  return {
    skill_ref: await descriptorSkillRef(descriptor),
    name: descriptor.name || "",
    description: descriptor.description || "",
    branch: "SKILL_BRANCH_MAIN",
    digest: descriptor.digest || "",
    manifest,
    manifest_truncated: manifestTruncated,
    entries,
    files: await Promise.all(paths.map((relPath) => focusedLocalFile(fs, path, root, relPath, maxFileChars))),
  };
}

function manifestCharLimit(args: Record<string, unknown>): number {
  const raw = args.max_manifest_chars ?? DEFAULT_FOCUS_MANIFEST_CHARS;
  const value = Number(raw);
  return Number.isFinite(value) ? value : DEFAULT_FOCUS_MANIFEST_CHARS;
}

function fileCharLimit(args: Record<string, unknown>): number {
  const raw = args.max_file_chars ?? DEFAULT_FOCUS_FILE_CHARS;
  const value = Number(raw);
  return Number.isFinite(value) ? value : DEFAULT_FOCUS_FILE_CHARS;
}

function pathsArg(item: Record<string, unknown>): string[] {
  if (!Array.isArray(item.paths)) {
    return [];
  }
  return item.paths
    .filter((value): value is string => typeof value === "string")
    .map((value) => value.trim())
    .filter(Boolean);
}

function includeManifest(item: Record<string, unknown>): boolean {
  return typeof item.include_manifest === "boolean" ? item.include_manifest : true;
}

async function focusedLocalFile(
  fs: typeof import("node:fs/promises"),
  path: typeof import("node:path"),
  root: string,
  relPath: string,
  maxFileChars: number,
): Promise<Record<string, unknown>> {
  const base = { path: relPath, branch: "SKILL_BRANCH_MAIN", content: "", truncated: false, size: 0 };
  const { TextDecoder } = await import("node:util");
  try {
    const full = path.resolve(root, relPath);
    const relative = path.relative(root, full);
    if (relative.startsWith("..") || path.isAbsolute(relative)) {
      return { ...base, error: { type: "skill_error", code: "invalid_skill_file_path", message: "path must stay inside the local skill root" } };
    }
    const stat = await fs.stat(full);
    if (!stat.isFile()) {
      return { ...base, error: { type: "skill_error", code: "skill_file_not_found", message: "skill file was not found" } };
    }
    const raw = await fs.readFile(full);
    let content = new TextDecoder("utf-8", { fatal: true }).decode(raw);
    let truncated = false;
    const chars = Array.from(content);
    if (maxFileChars > 0 && chars.length > maxFileChars) {
      content = chars.slice(0, maxFileChars).join("");
      truncated = true;
    }
    return { ...base, content, truncated, size: stat.size };
  } catch (error: any) {
    if (error?.code === "ENOENT") {
      return { ...base, error: { type: "skill_error", code: "skill_file_not_found", message: "skill file was not found" } };
    }
    if (error instanceof TypeError) {
      return { ...base, error: { type: "skill_error", code: "invalid_skill_file_utf8", message: "skill file must be valid UTF-8" } };
    }
    return { ...base, error: { type: "skill_error", code: "skill_file_read_failed", message: error instanceof Error ? error.message : String(error) } };
  }
}

async function walkFiles(fs: any, path: any, root: string, dir: string): Promise<string[]> {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  const out: string[] = [];
  for (const entry of entries) {
    if (entry.name === ".git" || entry.name === "node_modules") {
      continue;
    }
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      out.push(...await walkFiles(fs, path, root, full));
      continue;
    }
    if (entry.isFile()) {
      out.push(path.relative(root, full).split(path.sep).join("/"));
    }
  }
  return out;
}

async function localSkillsByRef(localSkills: LocalSkillDescriptor[]): Promise<Map<string, LocalSkillDescriptor>> {
  const out = new Map<string, LocalSkillDescriptor>();
  for (const skill of localSkills) {
    out.set(await descriptorSkillRef(skill), skill);
  }
  return out;
}

async function descriptorSkillRef(descriptor: LocalSkillDescriptor): Promise<string> {
  const existing = String(descriptor.skill_ref || "").trim();
  if (existing) {
    return existing;
  }
  return skillRefForLocal(descriptor.local_skill_id || "", descriptor.digest || "");
}

function skillRefForLocal(
  localSkillID: string,
  digest: string,
): string {
  const slug = skillRefSlug(localSkillID) || "local-skill";
  const digestPart = skillRefDigestPart(digest) || "unknown";
  return `local::${slug}@${digestPart}::main`;
}

function skillRefSlug(raw: string): string {
  return raw
    .trim()
    .toLowerCase()
    .replace(/::/g, "-")
    .replace(/@/g, "-")
    .replace(/[^a-z0-9_-]+/g, "-")
    .replace(/^[-_]+|[-_]+$/g, "")
    .slice(0, 64)
    .replace(/^[-_]+|[-_]+$/g, "");
}

function skillRefDigestPart(raw: string): string {
  const digest = raw.trim().toLowerCase().split(":").pop() || "";
  return digest.replace(/[^a-z0-9_-]+/g, "").slice(0, 16);
}

async function readLocalManifest(
  fs: typeof import("node:fs/promises"),
  path: typeof import("node:path"),
  root: string,
  maxManifestChars: number,
): Promise<{ manifest: string; manifestTruncated: boolean }> {
  const { TextDecoder } = await import("node:util");
  try {
    const raw = await fs.readFile(path.join(root, "SKILL.md"));
    let manifest = new TextDecoder("utf-8", { fatal: true }).decode(raw);
    let manifestTruncated = false;
    const chars = Array.from(manifest);
    if (maxManifestChars > 0 && chars.length > maxManifestChars) {
      manifest = chars.slice(0, maxManifestChars).join("");
      manifestTruncated = true;
    }
    return { manifest, manifestTruncated };
  } catch (error: any) {
    if (error?.code === "ENOENT") {
      return { manifest: "", manifestTruncated: false };
    }
    throw error;
  }
}

File: python/src/agent_api/local_skills.py
from __future__ import annotations

import hashlib
import json
import re
from pathlib import Path
from typing import Any

from agent_api.local_functions import function_call_output_input, pending_function_calls
from agent_api.types import AgentResponse, FunctionCallOutputInput
from agent_api.types.skills import LocalSkillDescriptor

DEFAULT_FOCUS_MANIFEST_CHARS = 16000
DEFAULT_FOCUS_FILE_CHARS = 12000


def local_skill_from_directory(
    root_dir: str | Path,
    *,
    id: str | None = None,
    name: str | None = None,
    description: str | None = None,
    max_manifest_chars: int = DEFAULT_FOCUS_MANIFEST_CHARS,
    metadata: dict[str, Any] | None = None,
) -> LocalSkillDescriptor:
    root = Path(root_dir).expanduser().resolve()
    digest = _directory_digest(root)
    local_id = id or root.name
    manifest, manifest_truncated = _read_local_manifest(root, max_manifest_chars)
    return {
        "local_skill_id": local_id,
        "skill_ref": _skill_ref_for_local(local_id, digest),
        "name": name or root.name,
        "description": description or "",
        "root_hint": str(root),
        "digest": digest,
        "manifest": manifest,
        "manifest_truncated": manifest_truncated,
        "metadata": metadata or {},
    }


def pending_local_skill_calls(response: AgentResponse) -> list[dict[str, Any]]:
    return [call for call in pending_function_calls(response) if call.get("name") == "skill_focus"]


async def run_local_skill_handlers(
    response: AgentResponse,
    local_skills: list[LocalSkillDescriptor],
) -> list[FunctionCallOutputInput]:
    outputs: list[FunctionCallOutputInput] = []
    by_ref = {_descriptor_skill_ref(skill): skill for skill in local_skills}
    for call in pending_local_skill_calls(response):
        raw_args = call.get("arguments") or "{}"
        args = json.loads(raw_args) if isinstance(raw_args, str) else dict(raw_args)
        payload = _focus_local_skills(args, by_ref)
        outputs.append(function_call_output_input(str(call.get("call_id", "")), payload))
    return outputs


def _focus_local_skills(args: dict[str, Any], by_ref: dict[str, LocalSkillDescriptor]) -> dict[str, Any]:
    max_manifest_chars = _manifest_char_limit(args)
    max_file_chars = _file_char_limit(args)
    items = args.get("skills") or []
    data: list[dict[str, Any]] = []
    if not isinstance(items, list):
        return {"data": [{"ok": False, "error": {"code": "invalid_skill_focus", "message": "skills must be an array"}}]}
    for item in items:
        if not isinstance(item, dict):
            data.append({"ok": False, "error": {"code": "invalid_skill_focus", "message": "skill item must be an object"}})
            continue
        skill_ref = str(item.get("skill_ref") or "").strip()
        result: dict[str, Any] = {"ok": False, "skill_ref": skill_ref, "branch": "main"}
        descriptor = by_ref.get(skill_ref)
        if descriptor is None:
            result["error"] = {"code": "skill_ref_not_found", "message": "skill_ref was not registered with the SDK"}
            data.append(result)
            continue
        try:
            result["skill"] = _focused_local_skill(descriptor, max_manifest_chars, _paths_arg(item), max_file_chars, _include_manifest(item))
            result["ok"] = True
        except Exception as exc:
            result["error"] = {"code": "local_skill_focus_failed", "message": str(exc)}
        data.append(result)
    return {"data": data}


def _focused_local_skill(descriptor: LocalSkillDescriptor, max_manifest_chars: int, paths: list[str], max_file_chars: int, include_manifest: bool) -> dict[str, Any]:
    root = Path(str(descriptor.get("root_hint", ""))).expanduser().resolve()
    if not root.is_dir():
        raise FileNotFoundError(f"local skill root does not exist: {root}")
    manifest_path = root / "SKILL.md"
    manifest = ""
    manifest_truncated = False
    if include_manifest and manifest_path.is_file():
        manifest = manifest_path.read_text(encoding="utf-8")
        if max_manifest_chars > 0 and len(manifest) > max_manifest_chars:
            manifest = manifest[:max_manifest_chars]
            manifest_truncated = True
    entries = []
    for path in sorted(root.iterdir(), key=lambda p: p.name):
        if path.name in {".git", "__pycache__", "node_modules"}:
            continue
        stat = path.stat()
        entries.append(
            {
                "path": path.name,
                "is_dir": path.is_dir(),
                "size": 0 if path.is_dir() else stat.st_size,
                "modified_at": int(stat.st_mtime),
            }
        )
    root_hint = str(root)
    return {
        "skill_ref": _descriptor_skill_ref(descriptor),
        "name": descriptor.get("name", ""),
        "description": descriptor.get("description", ""),
        "branch": "SKILL_BRANCH_MAIN",
        "digest": descriptor.get("digest", ""),
        "manifest": manifest,
        "manifest_truncated": manifest_truncated,
        "entries": entries,
        "files": [_focused_local_file(root, path, max_file_chars) for path in paths],
    }


def _manifest_char_limit(args: dict[str, Any]) -> int:
    raw = args.get("max_manifest_chars", DEFAULT_FOCUS_MANIFEST_CHARS)
    try:
        return int(raw)
    except (TypeError, ValueError):
        return DEFAULT_FOCUS_MANIFEST_CHARS


def _file_char_limit(args: dict[str, Any]) -> int:
    raw = args.get("max_file_chars", DEFAULT_FOCUS_FILE_CHARS)
    try:
        return int(raw)
    except (TypeError, ValueError):
        return DEFAULT_FOCUS_FILE_CHARS


def _paths_arg(item: dict[str, Any]) -> list[str]:
    raw = item.get("paths") or []
    if not isinstance(raw, list):
        return []
    return [str(path).strip() for path in raw if str(path).strip()]


def _include_manifest(item: dict[str, Any]) -> bool:
    raw = item.get("include_manifest")
    return raw if isinstance(raw, bool) else True


def _focused_local_file(root: Path, rel_path: str, max_file_chars: int) -> dict[str, Any]:
    result: dict[str, Any] = {"path": rel_path, "branch": "SKILL_BRANCH_MAIN", "content": "", "truncated": False, "size": 0}
    try:
        path = (root / rel_path).resolve()
        path.relative_to(root)
        if not path.is_file():
            result["error"] = {"type": "skill_error", "code": "skill_file_not_found", "message": "skill file was not found"}
            return result
        stat = path.stat()
        content = path.read_text(encoding="utf-8")
        truncated = False
        if max_file_chars > 0 and len(content) > max_file_chars:
            content = content[:max_file_chars]
            truncated = True
        result.update({"content": content, "truncated": truncated, "size": stat.st_size})
        return result
    except ValueError:
        result["error"] = {"type": "skill_error", "code": "invalid_skill_file_path", "message": "path must stay inside the local skill root"}
        return result
    except UnicodeDecodeError:
        result["error"] = {"type": "skill_error", "code": "invalid_skill_file_utf8", "message": "skill file must be valid UTF-8"}
        return result
    except Exception as exc:
        result["error"] = {"type": "skill_error", "code": "skill_file_read_failed", "message": str(exc)}
        return result


def _directory_digest(root: Path) -> str:
    h = hashlib.sha256()
    for path in sorted(p for p in root.rglob("*") if p.is_file() and ".git" not in p.parts and "__pycache__" not in p.parts):
        rel = path.relative_to(root).as_posix()
        h.update(rel.encode("utf-8"))
        h.update(b"\0")
        h.update(path.read_bytes())
        h.update(b"\0")
    return "sha256:" + h.hexdigest()


def _descriptor_skill_ref(descriptor: LocalSkillDescriptor) -> str:
    ref = str(descriptor.get("skill_ref") or "").strip()
    if ref:
        return ref
    return _skill_ref_for_local(str(descriptor.get("local_skill_id") or ""), str(descriptor.get("digest") or ""))


def _skill_ref_for_local(local_id: str, digest: str) -> str:
    slug = _skill_ref_slug(local_id) or "local-skill"
    digest_part = _skill_ref_digest_part(digest) or "unknown"
    return f"local::{slug}@{digest_part}::main"


def _skill_ref_slug(raw: str) -> str:
    slug = raw.strip().lower().replace("::", "-").replace("@", "-")
    slug = re.sub(r"[^a-z0-9_-]+", "-", slug).strip("-_")
    return slug[:64].strip("-_")


def _skill_ref_digest_part(raw: str) -> str:
    digest = raw.strip().lower().split(":")[-1]
    return re.sub(r"[^a-z0-9_-]+", "", digest)[:16]


def _read_local_manifest(root: Path, max_manifest_chars: int) -> tuple[str, bool]:
    manifest_path = root / "SKILL.md"
    if not manifest_path.is_file():
        return "", False
    manifest = manifest_path.read_text(encoding="utf-8")
    if max_manifest_chars > 0 and len(manifest) > max_manifest_chars:
        return manifest[:max_manifest_chars], True
    return manifest, False
